pmlc/dialect/stripe/transforms.cc

```diff
@@ -0,0 +1,205 @@
+//
+// Con ([outer_poly] + min(inner_poly) + const) >= 0 {
+//   PF [inner] {
+//     Pre
+//     GE Case
+//   }
+// } else {
+//   Cons ([outer_poly] + max(inner_poly) + const >= 0) {
+//     PF [inner] {
+//       Pre
+//       Con ([outer_poly] + [inner_poly] + const) >= 0 {
+//         GE Case
+//       } else {
+//         LT Case
+//       }
+//     }
+//   } else {
+//     PF [inner] {
+//       Pre
+//       LT Case
+//     }
+//   }
+// }
+//
+
+void LiftConstraint(ParallelForOp pf) {
+  // Extract the constraint op
+  auto pf_block = &pf.inner().front();
+  auto con = mlir::dyn_cast<ConstraintOp>(*std::prev(pf_block->end(), 2));
+  // Get the complete polynomial, and split it into innner + outer
+  auto opoly = AffinePolynomial(con.input());
+  AffinePolynomial ipoly;
+  for (mlir::BlockArgument* arg : pf_block->getArguments()) {
+    auto it = opoly.terms.find(arg);
+    if (it != opoly.terms.end()) {
+      ipoly.terms.emplace(*it);
+      opoly.terms.erase(it);
+    }
+  }
+  // Compute the range of the inner polynomial
+  auto irange = AffineRange(ipoly);
+  // Make a builder aimed right before the original parallel-for
+  OpBuilder builder(pf.getOperation());
+  Location loc = pf.getLoc();
+  // Build the rewritten version
+  auto oc = builder.create<ConstraintOp>(loc, builder.create<AffinePolyOp>(loc, opoly + AffinePolynomial(irange.min)));
+  // Make always true block
+  builder.createBlock(&oc.ge_case(), oc.ge_case().begin());
+  auto ge_pf = mlir::dyn_cast<ParallelForOp>(builder.clone(*pf.getOperation()));
+  builder.create<TerminateOp>(loc);
+  // Splice ge block out
+  ExtractConstraintCase(ge_pf, true);
+  builder.createBlock(&oc.lt_case(), oc.lt_case().begin());
+  auto ic = builder.create<ConstraintOp>(loc, builder.create<AffinePolyOp>(loc, opoly + AffinePolynomial(irange.max)));
+  builder.createBlock(&ic.ge_case(), ic.ge_case().begin());
+  builder.clone(*pf.getOperation());
+  builder.create<TerminateOp>(loc);
+  if (!con.lt_case().empty()) {
+    builder.createBlock(&ic.lt_case(), ic.lt_case().begin());
+    auto lt_pf = mlir::dyn_cast<ParallelForOp>(builder.clone(*pf.getOperation()));
+    builder.create<TerminateOp>(loc);
+    ExtractConstraintCase(lt_pf, true);
+  }
+  builder.setInsertionPointToEnd(&oc.lt_case().front());
+  builder.create<TerminateOp>(loc);
+  pf.getOperation()->erase();
+}
+
+bool SplitFor(ParallelForOp pf) {
+  if (!SafeConstraintInterior(pf)) {
+    return false;
+  }
+  // Extract the constraint op + poly
+  auto pf_block = &pf.inner().front();
+  auto con = mlir::dyn_cast<ConstraintOp>(*std::prev(pf_block->end(), 2));
+  auto poly = AffinePolynomial(con.input());
+  if (poly.terms.size() != 1) {
+    return false;  // Only works for simple polynomials
+  }
+  mlir::BlockArgument* ba = poly.terms.begin()->first;
+  if (ba->getOwner() != pf_block) {
+    return false;  // Only works when poly is over PF argument
+  }
+  std::cerr << "FOUND CANDIDATE:\n";
+  pf.dump();
+  return true;
```

Comment:
Can we use an IVLOG here instead?

Reply:
Whoops, this was supposed to be removed...